src/database/keyvalue.rs

```diff
@@ -379,6 +379,77 @@ impl Database for Tree {
         })?
         .map_or(Ok(0), ivec_to_u32)
     }
+
+    fn del_spent_utxos(
+        &mut self,
+        spent_outpoints: Option<Vec<OutPoint>>,
+    ) -> Result<Vec<LocalUtxo>, Error> {
+        let mut deleted_utxos = vec![];
+        if let Some(spent_outpoints) = spent_outpoints {
+            for spent_outpoint in spent_outpoints {
+                if let Some(utxo) = self.del_utxo(&spent_outpoint)? {
+                    deleted_utxos.push(utxo);
+                }
+            }
+            Ok(deleted_utxos)
+        } else {
+            let all_spent_utxos: Vec<LocalUtxo> = self
+                .iter_utxos()?
+                .into_iter()
+                .filter(|utxo| utxo.is_spent)
+                .collect();
+
+            for spent_utxo in all_spent_utxos {
+                if let Some(utxo) = self.del_utxo(&spent_utxo.outpoint)? {
+                    deleted_utxos.push(utxo);
+                }
+            }
+            Ok(deleted_utxos)
+        }
+    }
+
+    fn del_spent_utxos_by_criteria(
+        &mut self,
+        criteria: DelCriteria,
+    ) -> Result<Vec<LocalUtxo>, Error> {
+        let spent_utxos: Vec<LocalUtxo> = self
+            .iter_utxos()?
+            .into_iter()
+            .filter(|utxo| utxo.is_spent)
+            .collect();
+        let spent_utxo_size = spent_utxos.len();
+        if let Some(threshold_size) = criteria.threshold_size {
+            // if the threshold size is smaller than current size, delete everything
+            if threshold_size < spent_utxo_size as u64 {
+                //delete all spent utxos
+                return self.del_spent_utxos(None);
```

Comment:
How about deleting only the excess utxos sorted by confirmations??

Reply:
Will deleting just the excess not mean they will have to do frequent cleanups?

Reply:
Will deleting just the excess not mean they will have to do frequent cleanups?